File: 1197-parsing-a-boolean-expression/1197-parsing-a-boolean-expression.java
class Solution {
    public boolean parseBoolExpr(String expression) {
        Stack<Character> st = new Stack<>();

        for(char c : expression.toCharArray()) {
            if(c == '(' || c == ',') continue;
            if(c == 't' || c == 'f' || c == '!' || c == '&' || c == '|') st.add(c);
            else if(c == ')') {
                boolean hasTrue = false, hasFalse = false;
                while(st.peek() != '!' && st.peek() != '&' && st.peek() != '|') {
                    char t = st.pop();
                    if(t == 't') hasTrue = true;
                    if(t == 'f') hasFalse = true;
                }

                char t = st.pop();
                if(t == '!') st.add(hasTrue ? 'f' : 't');
                else if(t == '&') st.add(hasFalse ? 'f' : 't');
                else st.add(hasTrue ? 't' : 'f');
            }
        }

        return st.peek() == 't';
    }
}

File: 1197-parsing-a-boolean-expression/1197-parsing-a-boolean-expression.cpp
class Solution {
public:
    bool parseBoolExpr(string expression) {
        stack<char> st;

        for(char c : expression) {
            if(c == ',' || c == '(') continue;
            if(c == 't' || c == 'f' || c == '!' || c == '&' || c == '|') st.push(c);
            else if(c == ')') {
                bool hasTrue = false, hasFalse = false;
                while(st.top() != '!' && st.top() != '&' && st.top() != '|') {
                    char t = st.top();
                    st.pop();
                    if(t == 't') hasTrue = true;
                    if(t == 'f') hasFalse = true;
                }

                char t = st.top();
                st.pop();
                if(t == '!') st.push(hasTrue ? 'f' : 't');
                else if(t == '&') st.push(hasFalse ? 'f' : 't');
                else st.push(hasTrue ? 't' : 'f');
            }
        }

        return st.top() == 't';
    }
};